Issue - 141 - Validate macro actions (#141)

attilacsanyi opened the pull request:
Not able to handle auto disable over the save button, just prevent the saving.

mondalaci commented:
Why can't you disable the save button?

attilacsanyi commented:
Because the save button is on a top level `macro-action-editor.component` and the changes are in the children components. If I set to disabled then no chance to set back to active again as children `macro` components not emit their state.

_I try to avoid a redesign where all children emit its change event so parent can listen to and update its state. The easy validation functionality was not kept in mind in the beginning._

mondalaci commented:
I understand, but the user experience is not as great as if disabled was used. Please use disabled and take your time to redesign.

attilacsanyi commented:
Refactored validation logic @mondalaci, so now the `Save` button immediately reflect the validation state of each macro tab component. I have tested tons of cases and fixed some existing bugs as well. I hope you cannot find any missed validation cases ;-) Thanks for the testing in advance.

mondalaci commented:
Thanks, this is so much better now! Please also make both the X and Y coordinates of the mouse move and mouse scroll actions mandatory, otherwise the following can happen.

![untitled](https://user-images.githubusercontent.com/300402/29746196-78b70318-8ad2-11e7-9974-2497ed91c5a2.png)

attilacsanyi commented:
Thanks for the testing, should work in this case as well by design, will check it again @mondalaci ;-)

attilacsanyi commented:
Hmm, this is working for me could you please provide me repro steps? @mondalaci 
By design i implemented this `"Any of the X, Y fields are empty"`. Thanks

attilacsanyi commented:
Should be ok for now, so disabled if `"Any of the X, Y fields are empty"` in case of move or scroll @mondalaci

mondalaci commented:
Thanks, looks good to me! Waiting for @ert78gb or @fjozsef to review.

ert78gb commented on packages/uhk-web/src/app/components/macro/action-editor/tab/delay/macro-delay.component.ts:
```diff
@@ -8,6 +8,7 @@ import {
 } from '@angular/core';
 
 import { DelayMacroAction } from '../../../../../config-serializer/config-items/macro-action';
+import { MacroBaseComponent } from './../macro-base.component';
```
pls remove ./ from begining of the import

ert78gb commented on packages/uhk-web/src/app/components/macro/action-editor/tab/mouse/macro-mouse.component.ts:
```diff
@@ -7,6 +7,7 @@ import {
     MacroSubAction
 } from '../../../../../config-serializer/config-items/macro-action';
 import { Tab } from '../../../../popover/tab';
+import { MacroBaseComponent } from './../macro-base.component';
```
pls remove ./ from beginning of the import

attilacsanyi replied:
thx.

ert78gb commented on packages/uhk-web/src/app/components/macro/action-editor/tab/key/macro-key.component.ts:
```diff
@@ -71,4 +73,9 @@ export class MacroKeyTabComponent implements OnInit {
 
+    isMacroValid = () => {
+        const keyMacroAction = this.getKeyMacroAction();
+        return keyMacroAction.scancode !== undefined || keyMacroAction.modifierMask !== undefined;
```
pls use the following condition `!!keyMacroAction.scancode || !!keyMacroAction.modifierMask;`

@mondalaci If I read well the specification
>Key macro action: Scancode None and no modifiers selected

then the condition is && instead of ||

mondalaci replied:
@ert78gb It must be disabled if scancode is none and no modifiers is selected. Thus, it's be valid (not disabled) if scancode is not none or any modifiers are selected. The current implementation is correct.

attilacsanyi replied:
thank you guys.

ert78gb commented on packages/uhk-web/src/app/components/macro/action-editor/tab/mouse/macro-mouse.component.ts:
```diff
@@ -120,4 +126,18 @@ export class MacroMouseTabComponent implements OnInit {
+            case MoveMouseMacroAction:
+            case ScrollMouseMacroAction:
+                const { x, y } = this.macroAction as MoveMouseMacroAction;
+                return x !== undefined && x !== null && y !== undefined && y !== null;
```
pls use `return !!x && !!y;`

attilacsanyi replied:
better to keep the longer version as your solution no allow to set `0` to x or y which should be valid. @ert78gb

ert78gb commented on packages/uhk-web/src/app/components/macro/action-editor/tab/mouse/macro-mouse.component.ts:
```diff
@@ -120,4 +126,18 @@ export class MacroMouseTabComponent implements OnInit {
+                return x !== undefined && x !== null && y !== undefined && y !== null;
+            case MouseButtonMacroAction:
+                const { mouseButtonsMask } = this.macroAction as MouseButtonMacroAction;
+                return mouseButtonsMask !== undefined && mouseButtonsMask !== 0;
```
pls use `return !!mouseButtonsMask;`

ert78gb commented on packages/uhk-web/src/app/components/macro/action-editor/tab/text/macro-text.component.ts:
```diff
@@ -9,31 +9,39 @@ import {
         this.macroAction.text = this.input.nativeElement.value;
     }
 
+    isMacroValid = () => this.input.nativeElement.value !== undefined && this.input.nativeElement.value.length > 0;
```
pls use `!!this.input.nativeElement.value;`
